R/variance_ratio.R

```diff
@@ -33,100 +33,162 @@
+#'
 #'  #Calculate the variance ratio and CIs for each replicate
-#'  
-#'  res_withinreplicates <- variance_ratio(knz_001d, time.var = "year", species.var = "species", 
+#'
+#'  res_withinreplicates <- variance_ratio(knz_001d, time.var = "year", species.var = "species",
 #'  abundance.var = "abundance", bootnumber = 1, replicate = "subplot", average.replicates = FALSE)
-variance_ratio <- function(df, time.var, species.var, abundance.var, bootnumber, replicate.var=NA,
-                        li=0.025, ui=0.975,  average.replicates=TRUE) {
-  
+variance_ratio <- function(df, time.var, species.var, abundance.var,
+                           bootnumber, replicate.var,
+                           average.replicates = TRUE, level = 0.95, ...) {
+
   # check to make sure abundance is numeric data
   check_numeric(df, time.var, abundance.var)
-   
+
   # if no replicates, calculate single variance ratio
-   if(is.na(replicate.var)) {
-        check_single_onerep(df, time.var, species.var)
-        VR <- variance_ratio_longformdata(df, time.var, species.var, abundance.var)
+  if  (missing(replicate.var)) {
+
+    check_single_onerep(df, time.var, species.var)
+    VR <- variance_ratio_longformdata(df, time.var, species.var, abundance.var)
+
+    ## null models
+
+    nullval <- cyclic_shift(df, time.var = time.var,
+                            species.var = species.var,
+                            abundance.var = abundance.var,
+                            method = variance_ratio_matrixdata,
+                            bootnumber = bootnumber)
+
+    nullout <- confint(nullval)
+
+    # bind actual value and CI
+    output <- cbind(nullout, VR)
+
+  } else {
+
+    # if multiple replicates, check all replicates have values
+    check_single(df, time.var, species.var, replicate.var)
+    ## if you give a replicate, it must be a factor. This gives users responsibility for order.
+    assertthat::assert_that(is.factor(df[[replicate.var]]))
+
+
+    # calculate average variance ratio across replicates
+    if (average.replicates == TRUE) {
+      ## check the data
+      check_multispp(df, species.var, replicate.var)
+
+      # split the dataset into replicates, according to replicate.var
+      df <- df[order(df[[replicate.var]]),]
+      X <- split(df, df[replicate.var])
+
+      VR <- mean(unlist(lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var)))
+
+      ##
+      null_list <- lapply(X, FUN = cyclic_shift,  time.var = time.var,
+                          species.var = species.var,
+                          abundance.var = abundance.var,
+                          method = variance_ratio_matrixdata,
+                          bootnumber = bootnumber)
+
+      ## confidence intervals
+      null_output <- lapply(null_list, `[[`, i = "out")
+
+      null_means <- vector(mode = "numeric", length = bootnumber)
+
+      for (B in 1:bootnumber) {
+        null_means[B] <- mean(vapply(null_output, `[[`, 1, i = B))
+      }
```

Comment:
this for loop is probably needless. better to do `do.call(rbind, null_output)` followed by `colMeans`?

But in general, am I doing this correctly? right now we do this:
* split df based on replicate.var
* cyclic shift each replicate _independently_, obtaining one null distribution of variance ratio per replicate
* calculating the average variance ratio for all replicates -- all of the first numbers of the null distribution in each replicate, then all the second numbers, then all the third numbers etc
* average, quantiles of this vector of means.

Did I interpret that correctly? I am worried that it is incorrect. Shouldn't the same cyclic shift be used within each replicate that are combined in the same null mean? in other words, if we have four years of data (1, 2, 3, 4) and two replicates (A and B), shouldn't each randomization be the same in both?
A: 3,4,1,2
B: 3,4,1,2

but right now it could be something like
A: 3,4,1,2
B: 2,3,4,1